gflownet/utils/batch.py

```diff
@@ -0,0 +1,175 @@
+from copy import deepcopy
+
+import numpy as np
+import torch
+
+from gflownet.utils.common import tbool, tfloat, tint, tlong
+
+
+class Batch:
+    """
+    Important note: one env should correspond to only one trajectory, all env_id should be unique
+    """
+
+    def __init__(self, loss, device, float):
+        self.loss = loss
+        self.device = device
+        self.float = float
+        self.envs = dict()
+        self.state = []
+        self.action = []
+        self.done = []
+        self.env_id = []
+        self.mask_invalid_actions_forward = []
+        self.mask_invalid_actions_backward = []
+        self.parents = []
+        self.parents_actions = []
+        self.step = []
+
+    def __len__(self):
+        return len(self.state)
+
+    def add_to_batch(self, envs, actions, valids, train=True):
+        for env in envs:
+            self.envs.update({env.id: env})
```

Comment:
I don't understand the need for this. @AlexandraVolokhova could you explain it please?

Reply:
Yes, you are right, we can do it in the beginning of the loop behind, there's no need for two loops here. My bad, I didn't code it optimally here